In [ ]:
import pandas as pd
import seaborn as sb

df = sb.load_dataset("titanic")
df.head()


survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

In [ ]:
df.isnull().sum()

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

In [ ]:
# Fill missing values without inplace=True (recommended way)
df['age'] = df['age'].fillna(df['age'].median())
df['embarked'] = df['embarked'].fillna(df['embarked'].mode()[0])
df['embark_town'] = df['embark_town'].fillna(df['embark_town'].mode()[0])





In [ ]:
df.isnull().sum()

survived       0
pclass         0
sex            0
age            0
sibsp          0
parch          0
fare           0
embarked       0
class          0
who            0
adult_male     0
embark_town    0
alive          0
alone          0
dtype: int64

In [ ]:
df.select_dtypes(include=['number']).columns


Index(['survived', 'pclass', 'age', 'sibsp', 'parch', 'fare'], dtype='object')

In [ ]:
df.select_dtypes(exclude=['number']).columns


Index(['sex', 'embarked', 'class', 'who', 'adult_male', 'embark_town', 'alive',
       'alone'],
      dtype='object')

In [ ]:
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()

# Label Encoding for binary or yes/no type columns
binary_cols = ['sex', 'adult_male', 'alive', 'alone']
for col in binary_cols:
    df[col] = le.fit_transform(df[col])

# One-Hot Encoding for non-binary categorical columns
df = pd.get_dummies(df, columns=['embarked', 'class', 'who', 'embark_town'], drop_first=True)


KeyError: "None of [Index(['embarked', 'class', 'who', 'embark_town'], dtype='object')] are in the [columns]"

In [ ]:
df.columns


Index(['survived', 'pclass', 'sex', 'age', 'sibsp', 'parch', 'fare',
       'adult_male', 'alive', 'alone', 'embarked_1', 'embarked_2', 'who_1',
       'who_2', 'embark_town_1', 'embark_town_2'],
      dtype='object')

In [ ]:
df.head()

survived  pclass  sex   age  sibsp  parch     fare  adult_male  alive  \
0         0       3    1  22.0      1      0   7.2500           1      0   
1         1       1    0  38.0      1      0  71.2833           0      1   
2         1       3    0  26.0      0      0   7.9250           0      1   
3         1       1    0  35.0      1      0  53.1000           0      1   
4         0       3    1  35.0      0      0   8.0500           1      0   

   alone  embarked_1  embarked_2  who_1  who_2  embark_town_1  embark_town_2  
0      0       False        True   True  False          False           True  
1      0       False       False  False   True          False          False  
2      1       False        True  False   True          False           True  
3      0       False        True  False   True          False           True  
4      1       False        True   True  False          False           True

In [ ]:
from sklearn.preprocessing import StandardScaler, MinMaxScaler
scaler = StandardScaler()
num_cols = ['age', 'fare', 'sibsp', 'parch']  # choose your numeric columns except 'survived'
df[num_cols] = scaler.fit_transform(df[num_cols])


In [ ]:
df[num_cols].describe()


age          fare         sibsp         parch
count  8.910000e+02  8.910000e+02  8.910000e+02  8.910000e+02
mean  -2.791133e-17  3.987333e-18  1.594933e-17  5.283216e-17
std    1.000562e+00  1.000562e+00  1.000562e+00  1.000562e+00
min   -2.224156e+00 -6.484217e-01 -4.745452e-01 -4.736736e-01
25%   -5.657365e-01 -4.891482e-01 -4.745452e-01 -4.736736e-01
50%   -1.046374e-01 -3.573909e-01 -4.745452e-01 -4.736736e-01
75%    4.333115e-01 -2.424635e-02  4.327934e-01 -4.736736e-01
max    3.891554e+00  9.667167e+00  6.784163e+00  6.974147e+00

In [ ]:
from sklearn.preprocessing import StandardScaler, MinMaxScaler
scaler = MinMaxScaler()
num_cols = ['age', 'fare', 'sibsp', 'parch']  # choose your numeric columns except 'survived'
df[num_cols] = scaler.fit_transform(df[num_cols])



In [ ]:
df[num_cols].describe()

age        fare       sibsp       parch
count  891.000000  891.000000  891.000000  891.000000
mean     0.363679    0.062858    0.065376    0.063599
std      0.163605    0.096995    0.137843    0.134343
min      0.000000    0.000000    0.000000    0.000000
25%      0.271174    0.015440    0.000000    0.000000
50%      0.346569    0.028213    0.000000    0.000000
75%      0.434531    0.060508    0.125000    0.000000
max      1.000000    1.000000    1.000000    1.000000

In [ ]:
df.columns


Index(['survived', 'pclass', 'sex', 'age', 'sibsp', 'parch', 'fare',
       'adult_male', 'alive', 'alone', 'embarked_1', 'embarked_2', 'who_1',
       'who_2', 'embark_town_1', 'embark_town_2'],
      dtype='object')

In [ ]:
import pandas as pd

# Create a small sample DataFrame (just for learning)
sample_data = {
    'name': [
        'Braund, Mr. Owen Harris',
        'Cumings, Mrs. John Bradley (Florence Briggs Thayer)',
        'Heikkinen, Miss. Laina',
        'Allen, Mr. William Henry',
        'Moran, Mr. James',
        'McCarthy, Dr. Thomas',
        'Reynolds, Rev. George',
        'Palsson, Master. Gosta Leonard'
    ],
    'sibsp': [1, 1, 0, 0, 0, 0, 0, 3],
    'parch': [0, 0, 0, 0, 0, 0, 0, 1]
}

df = pd.DataFrame(sample_data)
df


name  sibsp  parch
0                            Braund, Mr. Owen Harris      1      0
1  Cumings, Mrs. John Bradley (Florence Briggs Th...      1      0
2                             Heikkinen, Miss. Laina      0      0
3                           Allen, Mr. William Henry      0      0
4                                   Moran, Mr. James      0      0
5                               McCarthy, Dr. Thomas      0      0
6                              Reynolds, Rev. George      0      0
7                     Palsson, Master. Gosta Leonard      3      1

In [ ]:
df.columns

Index(['name', 'sibsp', 'parch'], dtype='object')

In [ ]:
df['Title'] = df['name'].str.extract(' ([A-Za-z]+).', expand=False)
df['Title']


0        Mr
1       Mrs
2      Miss
3        Mr
4        Mr
5        Dr
6       Rev
7    Master
Name: Title, dtype: object

In [ ]:
df['Title'] = df['Title'].replace(['Mlle', 'Ms'], 'Miss')
df['Title'] = df['Title'].replace(['Mme'], 'Mrs')
df['Title'] = df['Title'].replace(['Dr', 'Rev', 'Major', 'Col', 'Capt', 'Sir', 'Don', 'Jonkheer', 'Lady', 'Countess'], 'Rare')


In [ ]:
df['Title'].value_counts()


Title
Mr        3
Rare      2
Mrs       1
Miss      1
Master    1
Name: count, dtype: int64

In [ ]:
df['FamilySize'] = df['sibsp'] + df['parch'] + 1
df[['name', 'sibsp', 'parch', 'FamilySize']]


name  sibsp  parch  FamilySize
0                            Braund, Mr. Owen Harris      1      0           2
1  Cumings, Mrs. John Bradley (Florence Briggs Th...      1      0           2
2                             Heikkinen, Miss. Laina      0      0           1
3                           Allen, Mr. William Henry      0      0           1
4                                   Moran, Mr. James      0      0           1
5                               McCarthy, Dr. Thomas      0      0           1
6                              Reynolds, Rev. George      0      0           1
7                     Palsson, Master. Gosta Leonard      3      1           5

In [ ]:
df['FamilySize'].mean()

np.float64(1.75)

In [ ]:
df.to_csv('titanic_preprocessed.csv', index=False)


# Titanic Dataset Preprocessing and Feature Engineering

## Missing Value Imputation
- Numerical columns: Filled missing values with median.
- Categorical columns: Filled missing values with mode.

## Encoding
- Applied One-Hot Encoding to categorical features such as 'embarked', 'who', 'embark_town'.

## Scaling
- Used StandardScaler and MinMaxScaler to normalize numerical columns.

## Feature Engineering
- Extracted 'Title' from the 'Name' column.
- Grouped similar/rare titles (Dr, Rev → Rare).
- Created new feature 'FamilySize' = SibSp + Parch + 1.

## Results
- Unique Titles: Mr, Rare, Mrs, Miss, Master
- Average FamilySize: 1.75
- Final dataset saved as `titanic_preprocessed.csv`
